File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] arr, int target) {
        int peak = findPeak(arr);
        int ans = binarySearch(arr, target, 0, peak);
        if(ans == -1) return binarySearch(arr, target, peak + 1, arr.length - 1);
        return ans;
    }
    public int findPeak(int[] arr){
    int start = 0, end = arr.length - 1;
    while(start <= end){
        int mid = start + (end - start) / 2;
        if( mid < end && arr[mid] > arr[mid + 1]) return mid;
        else if ( mid > start && arr[mid] < arr[mid - 1]) return mid - 1;
        else if (arr[start] >= arr[mid]) end = mid - 1;
        else if (arr[start] <= arr[mid]) start = mid + 1;
    }
    return -1;
}
    public int binarySearch(int[] arr, int target, int start, int end){
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(arr[mid] > target) end = mid - 1;
            else if(arr[mid] < target) start = mid + 1;
            else return mid;
         }
         return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.c
int findPivot(int *arr, int n){
    if(n == 1) return -1;
    for(int i = 0; i < n - 1; i++) if(arr[i] > arr[i+1]) return i;
    return -1;
}
int bs(int *arr, int start, int end, int target){
    while(start <= end){
        int mid = start + (end - start) / 2;
        if(arr[mid] == target) return mid;
        else if(arr[mid] > target) end = mid - 1;
        else start = mid + 1;
    }
    return -1;
}
int search(int* arr, int n, int target) {
    int piv = findPivot(arr, n);
    printf("%d", piv);
    if(piv == -1) return bs(arr, 0, n-1, target);
    if(piv == 0 && target == arr[piv]) return piv;
    if(piv == 0 && n > 1) return bs(arr, 1, n-1, target);
    else{
        int ans = bs(arr, 0, piv, target);
        if(ans != -1) return ans;
        else return bs(arr, piv + 1, n - 1, target);
    }
    return 0;
}